File: Source/LuaIntegration/Private/Subsystem/LuaGameInstanceSubsystem.cpp
// Dog Fight Game Code By CYM.


#include "Subsystem/LuaGameInstanceSubsystem.h"

#include "GameInstance/LuaGameInstance.h"

void ULuaGameInstanceSubsystem::Initialize(FSubsystemCollectionBase& Collection)
{
	Super::Initialize(Collection);

	// Invoke script initialize
	ScriptInit();
}

void ULuaGameInstanceSubsystem::Deinitialize()
{
	Super::Deinitialize();

	// Invoke script deinitialize
	ScriptDeinit();
}

ULuaGameInstance* ULuaGameInstanceSubsystem::GetLuaGameInstance() const
{
	return Cast<ULuaGameInstance>(GetGameInstance());
}

File: Content/Script/DogFight/Subsystem/LuaGameInstanceSubsystem.lua
---@class LuaGameInstanceSubsystem : ULuaGameInstanceSubsystem
local LuaGameInstanceSubsystem = UnrealClass()

function LuaGameInstanceSubsystem:ScriptInit()
    ---@type LuaGameInstance
    local LuaGameInstance = self:GetLuaGameInstance()
    if LuaGameInstance then
        LuaGameInstance:RegisterSubsystem(self)
    end
end

function LuaGameInstanceSubsystem:ScriptDeinit()
    ---@type LuaGameInstance
    local LuaGameInstance = self:GetLuaGameInstance()
    if LuaGameInstance then
        LuaGameInstance:UnregisterSubsystem(self)
    end
end

return LuaGameInstanceSubsystem